datastore/src/main/java/org/jboss/pnc/datastore/repositories/internal/AbstractRepository.java

```diff
@@ -18,116 +18,168 @@
 
     @Inject
     protected EntityManager entityManager;
 
-    protected JpaRepository<T, ID> springRepository;
-
-    protected JpaSpecificationExecutor<T> springSpecificationsExecutor;
-
-    public AbstractRepository() {
-    }
+    protected final Class<T> entityClass;
+    protected final Class<ID> idClass;
 
-    public AbstractRepository(
-            JpaRepository<T, ID> springRepository,
-            JpaSpecificationExecutor<T> springSpecificationsExecutor) {
-        this.springRepository = springRepository;
-        this.springSpecificationsExecutor = springSpecificationsExecutor;
+    protected AbstractRepository(Class<T> entityClass, Class<ID> idClass) {
+        this.entityClass = entityClass;
+        this.idClass = idClass;
     }
 
     @Override
     public T save(T entity) {
-        return springRepository.save(entity);
+        if (isNew(entity)) {
+            entityManager.persist(entity);
+            return entity;
+        } else {
+            return entityManager.merge(entity);
+        }
     }
 
     @Override
     public void delete(ID id) {
-        springRepository.deleteById(id);
+        Objects.requireNonNull(id, "The given id must not be null!");
+        T entity = this.queryById(id);
+        if (entity != null) {
+            this.delete(entity);
+        }
+    }
+
+    @Override
+    public void delete(T entity) {
+        Objects.requireNonNull(entity, "The entity must not be null!");
+        if (isNew(entity)) {
+            return;
+        }
+
+        T existing = (T) entityManager.find(entityClass, entity.getId());
+
+        // if the entity to be deleted doesn't exist, delete is a NOOP
+        if (existing == null) {
+            return;
+        }
+
+        entityManager.remove(entityManager.contains(entity) ? entity : entityManager.merge(entity));
     }
 
     @Override
     public void flushAndRefresh(T entity) {
-        springRepository.flush();
+        entityManager.flush();
         entityManager.refresh(entity);
     }
 
     @Override
     public List<T> queryAll() {
-        return springRepository.findAll();
+        return getQuery().getResultList();
     }
 
     @Override
-    public List<T> queryAll(PageInfo pageInfo, SortInfo sortInfo) {
-        return springRepository.findAll(PageableMapper.map(pageInfo, sortInfo)).getContent();
+    public List<T> queryAll(PageInfo pageInfo, SortInfo<T> sortInfo) {
+        return findAll(pageInfo, sortInfo);
     }
 
     @Override
     public T queryById(ID id) {
-        return springRepository.findById(id).orElse(null);
+        Objects.requireNonNull(id, "The given id must not be null!");
+        return this.entityManager.find(entityClass, id);
     }
 
     @Override
     public T queryByPredicates(Predicate<T>... predicates) {
-        return springSpecificationsExecutor.findOne(SpecificationsMapper.map(predicates)).orElse(null);
+        try {
+            return getQuery(null, predicates).setMaxResults(2).getSingleResult();
```

Comment:
?? why is this setMaxResults 2 if you want a single result?? I guess is copy-paste but wtf?

Reply:
I'ts copy pasta. How I understand is that having no maxResult, it could query without "LIMIT" in the SQL, potentionaly returning a lot of data from DB, before failing in java with NonUniqueResultException. Having maxResult 1 would always return at most 1 entity and never failing with NonUniqueResultException.

Reply:
https://github.com/spring-projects/spring-data-jpa/issues/2594
https://github.com/spring-projects/spring-data-jpa/commit/fa41595818eeb0d1da7f271a7f0554eb544fc6e3

Reply:
I see thx